src/btree/bt_sync.c

```diff
@@ -377,18 +422,32 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
         if (is_las) {
             LF_CLR(WT_READ_CACHE);
             LF_SET(WT_READ_CACHE_LEAF);
+            memset(&ref_list, 0, sizeof(ref_list));
         }
 
         for (;;) {
             WT_ERR(__sync_dup_walk(session, walk, flags, &prev));
             WT_ERR(__wt_tree_walk(session, &walk, flags));
 
-            if (walk == NULL)
+            if (walk == NULL) {
+                if (is_las) {
+                    for (i = 0; i < ref_list.entry; i++) {
+                        WT_IGNORE_RET_BOOL(__wt_page_evict_urgent(session, ref_list.list[i]));
+                        WT_ERR(__wt_page_release(session, ref_list.list[i], flags));
+                        WT_STAT_CONN_INCR(session, hs_gc_pages_evict);
+                        __wt_verbose(session, WT_VERB_CHECKPOINT_GC,
+                          "%p: is an in-memory obsolete page, added to urgent eviction queue.",
+                          (void *)ref_list.list[i]);
+                    }
+                    __sync_ref_list_free(session, &ref_list);
+                }
```

Comment:
I would suggest merging the urgent eviction setup work into a renamed `__sync_ref_list_free` call, instead of doing it inline (that is, we have a `__sync_ref_list_add()`, it's reasonable to have a corresponding `__sync_ref_list_pop()`).

But, more importantly, this could be a large number of pages, and there's no  point in accumulating all of them? Why not queue them up for eviction as soon as the checkpoint has passed by them? If there's cache pressure, that might help, and it can't hurt? And then the array doesn't just keep growing.

Reply:
I added a new function `__sync_ref_list_pop()` and updated according to your comment.

Regarding adding the pages to eviction queue intermittently may not solve the problem of eviction failure.  During the detection of this eviction fail issue, I found that eviction fails because of two reasons
1. Checkpoint holding the reference on the evicted page
2. Checkpoint holding the reference on the parent page also.

I doubt that adding pages to eviction queue intermittently may hit the second problem?

Reply:
What's the stack where holding a reference to the parent page prevents leaf page eviction?

Reply:
The failure can occur in the following code flow of eviction that checks the parent page is not reconciled by the checkpoint during the eviction writing a split page.

`__split_multi_lock()` -> `__split_internal_lock()` -> `__wt_btree_can_evict_dirty()`

Reply:
OK, you said "holding a reference", but it looks to me like this is a check of `WT_BTREE.sync`, that is, it's a test of whether or not the tree is currently being checkpointed and the thread isn't the thread performing the checkpoint.

So, yes, I'd agree with you that's going to be a problem.

I guess this pulls me back to my original question from awhile ago: we have pages from the LAS file we want to evict, and we have pages from the checkpoint that were pulled into the cache, and for someone reason we treat them differently. Do you understand why that is? In other words, why do we put one set on the urgent eviction queue, and the other set checkpoint itself goes and evicts.

I wonder if we could flag the page as being scheduled for eviction by the checkpointing thread, in order to allow this check to pass? We have a checkpoint generation for the overall checkpoint and for the file itself.